core/trino-main/src/main/java/io/trino/execution/scheduler/SqlQueryScheduler.java

```diff
@@ -173,69 +188,242 @@ private SqlQueryScheduler(
+                queryExecutor,
+                schedulerStats,
+                plan,
+                summarizeTaskInfo);
+        rootStageId = getStageId(queryStateMachine.getQueryId(), plan.getFragment().getId());
+        stageLineage = getStageLineage(queryStateMachine.getQueryId(), plan);
+        remotelyScheduledStages = stages.values().stream()
+                .filter(stage -> !stage.getFragment().getPartitioning().isCoordinatorOnly())
+                .map(SqlStageExecution::getStageId)
+                .collect(toImmutableSet());
+    }
+
+    // this is a separate method to ensure that the `this` reference is not leaked during construction
+    private void initialize()
+    {
+        // when query is done or any time a stage completes, attempt to transition query to "final query info ready"
+        queryStateMachine.addStateChangeListener(newState -> {
+            if (newState.isDone()) {
+                stages.values().forEach(SqlStageExecution::transitionToFinished);
+                queryStateMachine.updateQueryInfo(Optional.ofNullable(getStageInfo()));
+            }
+        });
+        for (SqlStageExecution stage : stages.values()) {
+            stage.addFinalStageInfoListener(status -> queryStateMachine.updateQueryInfo(Optional.ofNullable(getStageInfo())));
+        }
+    }
+
+    private static Map<PlanFragmentId, SqlStageExecution> createStages(
+            Session session,
+            RemoteTaskFactory taskFactory,
+            NodeTaskMap nodeTaskMap,
+            ExecutorService executor,
+            SplitSchedulerStats schedulerStats,
+            StageExecutionPlan planTree,
+            boolean summarizeTaskInfo)
+    {
+        ImmutableMap.Builder<PlanFragmentId, SqlStageExecution> result = ImmutableMap.builder();
+        for (StageExecutionPlan planNode : Traverser.forTree(StageExecutionPlan::getSubStages).breadthFirst(planTree)) {
+            PlanFragment fragment = planNode.getFragment();
+            SqlStageExecution stageExecution = createSqlStageExecution(
+                    getStageId(session.getQueryId(), fragment.getId()),
+                    fragment,
+                    planNode.getTables(),
+                    taskFactory,
+                    session,
+                    summarizeTaskInfo,
+                    nodeTaskMap,
+                    executor,
+                    schedulerStats);
+            result.put(fragment.getId(), stageExecution);
+        }
+        return result.build();
+    }
 
-        SqlStageExecution rootStage = stages.get(0);
-        rootStage.setOutputBuffers(rootOutputBuffers);
-        this.rootStageId = rootStage.getStageId();
+    private static Map<StageId, Set<StageId>> getStageLineage(QueryId queryId, StageExecutionPlan planTree)
+    {
+        ImmutableMap.Builder<StageId, Set<StageId>> result = ImmutableMap.builder();
+        for (StageExecutionPlan planNode : Traverser.forTree(StageExecutionPlan::getSubStages).breadthFirst(planTree)) {
```

Comment:
nit: name `planNode` source

Reply:
The method is refactored in the next commit